src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/IPersistentStorageLocationService.cs

```diff
@@ -53,11 +24,36 @@ public DefaultPersistentStorageLocationService()
         {
         }
 
-        public bool IsSupported(Workspace workspace) => false;
+        public virtual bool IsSupported(Workspace workspace) => false;
 
-        public string TryGetStorageLocation(SolutionId solutionId) => null;
-#pragma warning disable CS0067 // the event is unused
-        public event EventHandler<PersistentStorageLocationChangingEventArgs> StorageLocationChanging;
-#pragma warning disable CS0067
+        public string? TryGetStorageLocation(Solution solution)
+        {
+            if (!IsSupported(solution.Workspace))
+                return null;
+
+            if (string.IsNullOrWhiteSpace(solution.FilePath))
+                return null;
+
+            // Ensure that each unique workspace kind for any given solution has a unique
+            // folder to store their data in.
+
+            // Store in the LocalApplicationData/Roslyn/hash folder (%appdatalocal%/... on Windows,
+            // ~/.local/share/... on unix).  This will place the folder in a location we can trust
+            // to be able to get back to consistently as long as we're working with the same
+            // solution and the same workspace kind.
+            var appDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.Create);
+            var kind = StripInvalidPathChars(solution.Workspace.Kind ?? "");
+            var hash = StripInvalidPathChars(Checksum.Create(solution.FilePath).ToString());
+
+            return Path.Combine(appDataFolder, "Roslyn", "Cache", kind, hash);
```

Comment:
❓ What do we do if more than one copy of Visual Studio has the same solution open?

Reply:
Great question, and i can certainly add docs.  The system we've taken (which i'm not personally happy with) is that "first vs wins by grabbing an exclusive lock on the db".  so the second VS simply has no cache it can refer to.

Note: this was not in my original design when i did the sqlite work.  The original design was: we can have a single db instance can be shared among all running VS instances (as well as OOP).  Unfortunately, we discovered that one component that moved onto the sqlite persistence layer assumed otherwise.  i.e. it assumed it could just read/write from keys and there would be nothing else affecting those keys.

All the features i wrote specifically use a 'key+checksum' approach so that they can always validate if its safe to use the key/checksum/value they read in.  This other system did not, and so the work happening in OOP could corrupt it.  IMO, we should have just fixed the feature, but at the time for expediency it was just decided to lock the DB off, and to also make it so that OOP uses another db entirely.  This is unfortunately fairly wasteful, but i haven't ever had time to go back and fix that one internal feature to work properly on a key/value store that can be updated concurrently.